### ISSUE TITLE
Build failure: python3.11-azure-cli

### ISSUE THREAD

**Issue by khaneliman** [NixOS/nixpkgs#263510]
### Steps To Reproduce

1. build azure-cli

### Build log

```
error: builder for '/nix/store/qbk4jfwi7gyhfy5qdynwl52fsgh3flnm-python3.11-azure-cli-core-2.53.0.drv' failed with exit code 1;
       last 50 log lines:
       > azure/cli/core/tests/test_aaz_operation.py:402
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:402: DeprecationWarning: invalid escape sequence '\['
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Couldn't find 'o' in 'l\[0\]'. Available options: \['p', 's'\]"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:444
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:444: DeprecationWarning: invalid escape sequence '\['
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Couldn't find 't' in 'l\[1\]'. Available options: \['p', 's'\]"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:464
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:464: DeprecationWarning: invalid escape sequence '\('
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Expect <class 'str'>, got 3 \(<class 'int'>\)"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:484
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:484: DeprecationWarning: invalid escape sequence '\['
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Couldn't find '\[10\]' in 'd'. Available options: \['a', 'b', 'c'\]"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:495
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:495: DeprecationWarning: invalid escape sequence '\('
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Expect <class 'int'>, got a \(<class 'str'>\)"):
       >
       > azure/cli/core/extension/__init__.py:11
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/extension/__init__.py:11: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
       >     from distutils.sysconfig import get_python_lib  # pylint: disable=deprecated-module
       >
       > azure/cli/core/extension/__init__.py:11
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/extension/__init__.py:11: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
       >     from distutils.sysconfig import get_python_lib  # pylint: disable=deprecated-module
       >
       > azure/cli/core/tests/test_help_loaders.py:13
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_help_loaders.py:13: PytestCollectionWarning: cannot collect test class 'TestCommandLoader' because it has a __init__ constructor (from: azure/cli/core/tests/test_help_loaders.py)
       >     class TestCommandLoader(AzCommandsLoader):
       >
       > azure/cli/core/tests/test_application.py::TestApplication::test_client_request_id_is_refreshed_after_execution
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_application.py:56: DeprecationWarning: Please use assertNotEqual instead.
       >     self.assertNotEquals(old_id, cli.data['headers']['x-ms-client-request-id'])
       >
       > azure/cli/core/tests/test_application.py::TestApplication::test_client_request_id_is_refreshed_correctly
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_application.py:35: DeprecationWarning: Please use assertNotEqual instead.
       >     self.assertNotEquals(old_id, cli.data['headers']['x-ms-client-request-id'])
       >
       > azure/cli/core/tests/test_command_registration.py: 10 warnings
       >   /nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/lib/python3.11/unittest/case.py:1176: DeprecationWarning: assertDictContainsSubset is deprecated
       >     warnings.warn('assertDictContainsSubset is deprecated',
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED azure/cli/core/tests/test_command_registration.py::TestCommandRegistration::test_command_index - AttributeError: <MagicMock id='140737265805712'> does not have the attribut...
       > FAILED azure/cli/core/tests/test_command_registration.py::TestCommandRegistration::test_command_index_always_loaded_extensions - AssertionError: [] != 'FAKE_HANDLER'
       > FAILED azure/cli/core/tests/test_parser.py::TestParser::test_parser_error_spellchecker - AssertionError: 0 != 5
       > =========== 3 failed, 237 passed, 3 deselected, 22 warnings in 6.15s ===========
       For full logs, run 'nix log /nix/store/qbk4jfwi7gyhfy5qdynwl52fsgh3flnm-python3.11-azure-cli-core-2.53.0.drv'.
error: 1 dependencies of derivation '/nix/store/2fnf2n69hxkair57l0jm4fjxlsbn3s4a-python3.11-azure-cli-2.53.0.drv' failed to build

```

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
🔨 Welcome to khanelinix
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.90.1-microsoft-standard-WSL2, NixOS, 23.11 (Tapir), 23.11.20231019.7c9cc5a`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`

```

**Comment by mdarocha** [comment]
For now, changing python used to `python310` seems to fix the build:

```
azure-cli = super.azure-cli.override {
 python3 = super.python310;
};
```

**Comment by mdarocha** [comment]
Made a PR that fixes the failure.

**Comment by AkechiShiro** [comment]
Will investigate this failure and try to fix it (by using Python 3.11), but the build was working before (We may temporarily fix it using #263637 as a workaround), some dependencies must have been changed, from before, not sure which one is causing this failure, I'd need to bisect commits on master branch, I guess...

**Comment by mdarocha** [comment]
> Will investigate this failure and try to fix it (by using Python 3.11), but the build was working before (We may temporarily fix it using #263637 as a workaround), some dependencies must have been changed, from before, not sure which one is causing this failure, I'd need to bisect commits on master branch, I guess...

The failure is because of python3.11, the tests didn't work properly on 3.11. Upstream fixed those tests for a future release.

**Comment by AkechiShiro** [comment]
Thanks a lot for those clarifications and appreciating a lot your help in fixing all this @mdarocha 
However, we can't really make sure that those tests won't break again, unless we pin a patch on those ? And this is what you've done in your fix PR.

I'll be trying the latest release 2.53.1 with the dependencies cleanup.
